src/main/java/com/github/princesslana/eriscasper/repository/event/GuildsFromEvents.java

```diff
@@ -0,0 +1,93 @@
+package com.github.princesslana.eriscasper.repository.event;
+
+import com.github.princesslana.eriscasper.data.Snowflake;
+import com.github.princesslana.eriscasper.data.event.*;
+import com.github.princesslana.eriscasper.data.resource.Channel;
+import com.github.princesslana.eriscasper.data.resource.Guild;
+import com.github.princesslana.eriscasper.data.resource.ImmutableChannel;
+import com.github.princesslana.eriscasper.repository.GuildRepository;
+import io.reactivex.Maybe;
+import io.reactivex.Observable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Stack;
+
+public class GuildsFromEvents implements GuildRepository {
+
+  private final Map<Snowflake, Guild> guildCache;
+  private final Map<Snowflake, Channel> channelCache;
+
+  public GuildsFromEvents(Observable<Event> eventObservable) {
+    guildCache = new HashMap<>();
+    channelCache = new HashMap<>();
+    eventObservable
+        .ofType(GuildCreateEvent.class)
+        .map(GuildCreateEvent::unwrap)
+        .subscribe(
+            guild -> {
+              guildCache.putIfAbsent(guild.getId(), guild);
+              guild
+                  .getChannels()
+                  .forEach(
+                      channel ->
+                          channelCache.putIfAbsent(
+                              channel.getId(),
+                              ImmutableChannel.builder()
+                                  .from(channel)
+                                  .guildId(guild.getId())
+                                  .build()));
+            });
+    eventObservable
+        .ofType(GuildDeleteEvent.class)
+        .map(GuildDeleteEvent::unwrap)
+        .subscribe(
+            guild -> {
+              Snowflake id = guild.getId();
+              guildCache.remove(id);
+              Stack<Snowflake> removingChannelStack = new Stack<>();
+              channelCache
+                  .entrySet()
+                  .stream()
+                  .filter(entry -> id.equals(entry.getValue().getGuildId().orElse(null)))
+                  .forEach(entry -> removingChannelStack.push(entry.getKey()));
```

Comment:
Why do we push onto the stack here, rather than just deleting here?

Reply:
ConcurrentModificationException, I suppose switching to concurrent hashmaps should fix that so I'll remove the stack and see if tests still pass.